framework/src/argument.rs

```diff
@@ -190,7 +190,106 @@ pub async fn rest_argument_parse<T>(
     segments: &mut ArgumentSegments<'_>,
 ) -> Result<T, ArgumentError<T::Err>>
 where
-    T: serenity::utils::Parse,
+    T: Parse,
 {
     T::parse(ctx, msg, segments.source()).await.map_err(ArgumentError::Argument)
 }
+
+/// Denotes a type that can be either one of two different types.
+///
+/// It derives the [`Parse`] trait and can be used to parse an argument as either of two types.
+/// It attempts to parse into the type that is indicated first. If parsing into the first type fails,
+/// an attempt to parse into the second type is made. If both attempts fail, the overall parsing
+/// fails and returns a [`ParseEitherError`].
+///
+/// This can also be used to handle larger combinations of types by chaining [`ParseEither`]s,
+/// for example, `ParseEither<f32, ParseEither<i32, String>>`.
+#[derive(Debug)]
+#[non_exhaustive]
```

Comment:
I doubt we are going to expand upon this enumeration by adding new variants in the future. This just makes it annoying to `match` as you always have to include a default branch arm (`_ => ...`).
```suggestion
```